File: examples/athenademo/athenatest.py

from twisted.trial import unittest
from twisted.internet import defer

from nevow import athena, loaders, tags, util

TEST = 'return Nevow.Athena.Widget.get(this).test();'

class AthenaTestMixin:
    jsModules = {
        u'AthenaTest': util.resource_filename('athenademo', 'athenatest.js')}


class ImportBeforeLiteralJavaScript(AthenaTestMixin, athena.LiveFragment):

    jsClass = u'AthenaTest.ImportBeforeLiteralJavaScript'

    docFactory = loaders.stan(tags.p(render=tags.directive('liveFragment'))[
        tags.button(onclick=TEST)['Test Import Before Literal JavaScript'],
        tags.div[
            tags.script(type="text/javascript")["""
            var importBeforeLiteralJavaScriptResult;
            if (typeof Nevow.Athena.Widget == 'undefined') {
                importBeforeLiteralJavaScriptResult = false;
            } else {
                importBeforeLiteralJavaScriptResult = true;
            }
            """]]])



class ClientToServerArgumentSerialization(athena.LiveFragment, unittest.TestCase):
    """
    Tests that arguments passed to a method on the server are properly
    received.
    """

    jsClass = u'AthenaTest.ClientToServerArgumentSerialization'

    docFactory = loaders.stan(tags.div(render=tags.directive('liveFragment'))[
        tags.form(action='#', onsubmit=TEST)[
            tags.input(type='submit', value='Test Client To Server Argument Serialization')]])

    allowedMethods = {'test': True}
    def test(self, i, f, s, l, d):
        self.assertEquals(i, 1)
        self.assertEquals(f, 1.5)
        self.failUnless(isinstance(s, unicode))
        self.assertEquals(s, u'Hello world')
        self.failUnless(isinstance(l[2], unicode))
        self.assertEquals(l, [1, 1.5, u'Hello world'])
        self.assertEquals(d, {u'hello world': u'object value'})
        self.failUnless(isinstance(d.keys()[0], unicode))
        self.failUnless(isinstance(d.values()[0], unicode))



class ClientToServerResultSerialization(athena.LiveFragment):
    """
    Tests that the return value from a method on the server is
    properly received by the client.
    """

    jsClass = u'AthenaTest.ClientToServerResultSerialization'

    docFactory = loaders.stan(tags.div(render=tags.directive('liveFragment'))[
        tags.form(action='#', onsubmit=TEST)[
            tags.input(type='submit', value='Test Client To Server Result Serialization')]])

    allowedMethods = {'test': True}
    def test(self, i, f, s, l, d):
        return (i, f, s, l, d)


class ClientToServerExceptionResult(athena.LiveFragment):
    """
    Tests that when a method on the server raises an exception, the
    client properly receives an error.
    """

    jsClass = u'AthenaTest.ClientToServerExceptionResult'

    docFactory = loaders.stan(tags.div(render=tags.directive('liveFragment'))[
        tags.form(action='#', onsubmit='return Nevow.Athena.Widget.get(this).test(true);')[
            tags.input(type='submit', value='Test Client To Server Synchronous Exception Result')],
        tags.form(action='#', onsubmit='return Nevow.Athena.Widget.get(this).test(false);')[
            tags.input(type='submit', value='Test Client To Server Asynchronous Exception Result')]])


    allowedMethods = {'testSync': True, 'testAsync': True}
    def testSync(self, s):
        raise Exception(s)

    def testAsync(self, s):
        return defer.fail(Exception(s))


class ServerToClientArgumentSerialization(athena.LiveFragment):
    """
    Tests that a method invoked on the client by the server is passed
    the correct arguments.
    """

    jsClass = u'AthenaTest.ServerToClientArgumentSerialization'

    docFactory = loaders.stan(tags.div(render=tags.directive('liveFragment'))[
        tags.form(action='#', onsubmit=TEST)[
            tags.input(type='submit', value='Test Server To Client Argument Serialization')]])

    allowedMethods = {'test': True}
    def test(self):
        return self.callRemote('reverse', 1, 1.5, u'hello', {u'world': u'value'});

class ServerToClientResultSerialization(athena.LiveFragment, unittest.TestCase):
    """
    Tests that the result returned by a method invoked on the client
    by the server is correct.
    """

    jsClass = u'AthenaTest.ServerToClientResultSerialization'

    docFactory = loaders.stan(tags.div(render=tags.directive('liveFragment'))[
        tags.form(action='#', onsubmit=TEST)[
            tags.input(type='submit', value='Test Server To Client Result Serialization')]])

    allowedMethods = {'test': True}
    def test(self):
        def cbResults(result):
            self.assertEquals(result[0], 1)
            self.assertEquals(result[1], 1.5)
            self.assertEquals(result[2], u'hello')
            self.assertEquals(result[3], {u'world': u'value'})
        d = self.callRemote('reverse')
        d.addCallback(cbResults)
        return d



class WidgetInATable(AthenaTestMixin, athena.LiveFragment):
    jsClass = u"AthenaTest.WidgetInATable"

    docFactory = loaders.stan(tags.table[
        tags.tbody[
            tags.tr[
                tags.td(render=tags.directive('liveFragment'))[
                    tags.button(onclick=TEST)[
                        'Test Widget In A Table']]]]])



class WidgetIsATable(AthenaTestMixin, athena.LiveFragment):
    jsClass = u"AthenaTest.WidgetIsATable"

    docFactory = loaders.stan(tags.table(render=tags.directive('liveFragment'))[
        tags.tbody[
            tags.tr[
                tags.td[
                    tags.button(onclick=TEST)[
                        'Test Widget Is A Table']]]]])



class ParentChildRelationshipTest(AthenaTestMixin, athena.LiveFragment):
    jsClass = u"AthenaTest.ChildParentRelationshipTest"

    docFactory = loaders.stan(tags.div(render=tags.directive('liveFragment'))[
        tags.button(onclick=TEST)['Test Parent Child Relationships'],
        tags.div(render=tags.directive('childrenWidgets'))])


    def render_childrenWidgets(self, ctx, data):
        for i in xrange(3):
            yield ChildFragment(self.page, i)


    allowedMethods = {'getChildCount': True}
    def getChildCount(self):
        return 3



class ChildFragment(athena.LiveFragment):
    jsClass = u'AthenaTest.ChildParentRelationshipTest'

    docFactory = loaders.stan(tags.div(render=tags.directive('liveFragment'))[
        tags.div(render=tags.directive('childrenWidgets')),
        'child'])

    def __init__(self, page, childCount):
        super(ChildFragment, self).__init__()
        self.page = page
        self.childCount = childCount


    def render_childrenWidgets(self, ctx, data):
        # yield tags.div['There are ', self.childCount, 'children']
        for i in xrange(self.childCount):
            yield ChildFragment(self.page, self.childCount - 1)


    allowedMethods = {'getChildCount': True}
    def getChildCount(self):
        return self.childCount



class AutomaticClass(AthenaTestMixin, athena.LiveFragment):
    jsClass = u'AthenaTest.AutomaticClass'

    docFactory = loaders.stan(tags.div(render=tags.directive('liveFragment'))[
        tags.button(onclick=TEST)['Automatic athena:class attribute']])



class AthenaTests(athena.LivePage):
    docFactory = loaders.stan([
        tags.xml('<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">'),
        tags.html(**{'xmlns:nevow': 'http://nevow.com/ns/nevow/0.1'})[
            tags.head[
                tags.invisible(render=tags.directive('liveglue'))],
            tags.body[
                tags.slot('tests'),
                tags.invisible(render=tags.directive('introspection'))]]])

    addSlash = True

    tests = [
        # ImportBeforeLiteralJavascript _must_ be the first test.
        ImportBeforeLiteralJavaScript,

        ClientToServerArgumentSerialization,
        ClientToServerResultSerialization,
        ClientToServerExceptionResult,
        ServerToClientArgumentSerialization,
        ServerToClientResultSerialization,
        WidgetInATable,
        WidgetIsATable,
        AutomaticClass,
        ParentChildRelationshipTest,
        ]


    def renderTests(self):
        for frgClass in self.tests:
            frg = frgClass()
            self.jsModules.mapping.update(getattr(frg, 'jsModules', {}))
            frg.page = self
            yield frg


    def beforeRender(self, ctx):
        ctx.fillSlots('tests', self.renderTests())


    def render_introspection(self, ctx, data):
        f = athena.IntrospectionFragment()
        f.setFragmentParent(self)
        return ctx.tag[f]

File: examples/athenademo/athenatest.js

// import Nevow.Athena

AthenaTest = {};

AthenaTest.TestCase = Nevow.Athena.Widget.subclass('AthenaTest.TestCase');
AthenaTest.TestCase.methods(
    function fail(self, msg) {
        throw new Error('Test Failure: ' + msg);
    },

    function assertEquals(self, a, b) {
        if (!(a == b)) {
            fail(a + ' != ' + b);
        }
    },

    function test(self) {
        var res;
        var subargs = [];
        for (var i = 1; i < arguments.length; ++i) {
            subargs.push(arguments[i]);
        }
        try {
            res = self._test.apply(self, subargs);
        } catch (e) {
            res = MochiKit.Async.fail(e);
        }

        if (res == undefined || !res.addCallback || !res.addErrback) {
            res = MochiKit.Async.succeed(res);
        }

        res.addCallback(function(result) {
            Divmod.log('test', 'Success!');
        });

        res.addErrback(function(err) {
            Divmod.log('test', 'Failure: ' + err.message);
        });

        return false;
    });

AthenaTest.ClientToServerArgumentSerialization = AthenaTest.TestCase.subclass('AthenaTest.ClientToServerArgumentSerialization');
AthenaTest.ClientToServerArgumentSerialization.methods(
    function _test(self) {
        var L = [1, 1.5, 'Hello world'];
        var O = {'hello world': 'object value'};
        return self.callRemote('test', 1, 1.5, 'Hello world', L, O);
    });

AthenaTest.ClientToServerResultSerialization = AthenaTest.TestCase.subclass('AthenaTest.ClientToServerResultSerialization');
AthenaTest.ClientToServerResultSerialization.methods(
    function _test(self) {
        var L = [1, 1.5, 'Hello world'];
        var O = {'hello world': 'object value'};
        var d = self.callRemote('test', 1, 1.5, 'Hello world', L, O);
        d.addCallback(function(result) {
            self.assertEquals(result[0], 1);
            self.assertEquals(result[1], 1.5);
            self.assertEquals(result[2], 'Hello world');
            self.assertEquals(result[3][0], 1);
            self.assertEquals(result[3][1], 1.5);
            self.assertEquals(result[3][2], 'Hello world');
            self.assertEquals(result[4]['hello world'], 'object value');
        });
        return d;
    });

AthenaTest.ClientToServerExceptionResult = AthenaTest.TestCase.subclass('AthenaTest.ClientToServerExceptionResult');
AthenaTest.ClientToServerExceptionResult.methods(
    function _test(self, sync) {
        var d;
        var s = 'This exception should appear on the client.';
        if (sync) {
            d = self.callRemote('testSync', s);
        } else {
            d = self.callRemote('testAsync', s);
        }
        d.addCallbacks(
            function(result) {
                self.fail('Erroneously received a result: ' + result);
            },
            function(err) {
                var idx = err.message.indexOf(s);
                if (idx == -1) {
                    self.fail('Did not find expected message in error message: ' + err);
                }
            });
        return d;
    });

AthenaTest.ServerToClientArgumentSerialization = AthenaTest.TestCase.subclass('AthenaTest.ServerToClientArgumentSerialization');
AthenaTest.ServerToClientArgumentSerialization.methods(
    function _test(self) {
        return self.callRemote('test');
    },

    function reverse(self, i, f, s, o) {
        self.assertEquals(i, 1);
        self.assertEquals(f, 1.5);
        self.assertEquals(s, 'hello');
        self.assertEquals(o['world'], 'value');
    });

AthenaTest.ServerToClientResultSerialization = AthenaTest.TestCase.subclass('AthenaTest.ServerToClientResultSerialization');
AthenaTest.ServerToClientResultSerialization.methods(
    function _test(self) {
        return self.callRemote('test');
    },

    function reverse(self) {
        return [1, 1.5, 'hello', {'world': 'value'}];
    });

AthenaTest.WidgetInATable = AthenaTest.TestCase.subclass('AthenaTest.WidgetInATable');
AthenaTest.WidgetInATable.methods(
    function _test(self) {
        // Nothing to do
    });

AthenaTest.WidgetIsATable = AthenaTest.TestCase.subclass('AthenaTest.WidgetIsATable');
AthenaTest.WidgetIsATable.methods(
    function _test(self) {
        // Nothing to do
    });

AthenaTest.ChildParentRelationshipTest = AthenaTest.TestCase.subclass('AthenaTest.ChildParentRelationshipTest');
AthenaTest.ChildParentRelationshipTest.methods(
    function checkParent(self, proposedParent) {
        self.assertEquals(self.widgetParent, proposedParent);
    },

    function _test(self) {
        var deferredList = function(finalDeferred, counter) {
            if (counter == 0) {
                finalDeferred.callback(null);
            }
            var callback = function(ignored) {
                counter -= 1;
                Divmod.log('test', 'One more down, ' + counter + ' to go.');
                if (counter == 0) {
                    finalDeferred.callback(null);
                }
            };
            return callback;
        };

        var result = self.callRemote('getChildCount');

        result.addCallback(function(count) {
            Divmod.log('test', 'Discovered I have ' + count + ' children');
            var d = new MochiKit.Async.Deferred();
            d.addCallback(function() { self.node.style.border = 'thin solid green'; });
            var cb = deferredList(d, count);
            self.assertEquals(self.childWidgets.length, count);
            for (var i = 0; i < self.childWidgets.length; i++) {
                var childWidget = self.childWidgets[i];
                childWidget.checkParent(self);
                childWidget._test().addCallback(cb).addErrback(function(err) { d.errback(err); });
            }
            return d;
        });
        return result;
    });

AthenaTest.AutomaticClass = AthenaTest.TestCase.subclass('AthenaTest.AutomaticClass');
AthenaTest.AutomaticClass.methods(
    function _test(self) {
        // Nothing to do here
    });


AthenaTest.ImportBeforeLiteralJavaScript = AthenaTest.TestCase.subclass('AthenaTest.ImportBeforeLiteralJavaScript');
AthenaTest.ImportBeforeLiteralJavaScript.methods(
    function _test(self) {
        self.assertEquals(importBeforeLiteralJavaScriptResult, false);
    });
